bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoZoneHandler.java

```diff
@@ -403,9 +442,32 @@ private void scheduleHvacChange() {
         }, configuration.hvacChangeDebounce, TimeUnit.SECONDS);
     }
 
-    private void updateStateIfNotNull(String channelID, State state) {
-        if (state != null) {
-            updateState(channelID, state);
+    /**
+     * Helper method to log an API transaction on the given object.
+     * If the logger level is 'debug', the transaction is simply logged.
+     * If the logger level is 'trace, the object's JSON serial contents are included.
+     *
+     * @param object the object to be logged.
+     * @param isCommand marks whether the transaction is a command to, or a response from, the server.
+     */
+    private void logApiTransaction(Object object, boolean isCommand) {
+        if (logger.isDebugEnabled() || logger.isTraceEnabled()) {
+            String logType = isCommand ? "command" : "response";
+            if (logger.isDebugEnabled()) {
+                logger.debug("Api {}: homeId:{}, zoneId:{}, objectId:{}", logType, getHomeId(), getZoneId(),
+                        object.getClass().getSimpleName());
+            } else {
+                logger.trace("Api {}: homeId:{}, zoneId:{}, objectId:{}, content:\n{}", logType, getHomeId(),
+                        getZoneId(), object.getClass().getSimpleName(), convertToJsonString(object));
+            }
+        }
+    }
+
+    private String convertToJsonString(Object object) {
+        Gson gson = this.gson;
+        if (gson == null) {
+            gson = this.gson = GsonBuilderFactory.defaultGsonBuilder().setPrettyPrinting().create();
         }
```

Comment:
This is not thread safe, so if this method can be called in multithreaded context, you need some lock/mutex, for example by making the method `synchronized`.

Reply:
Ok. Good catch.